lib/browse_everything/driver/box.rb

```diff
@@ -10,103 +11,107 @@ def icon
+        results
       end
 
-      def link_for(path)
-        file = box_client.file(path)
+      # @param [String] id of the file in Box
+      # @return [Array<String, Hash>]
+      def link_for(id)
+        file = box_client.file_by_id(id)
         download_url = file.download_url
         auth_header = { 'Authorization' => "Bearer #{@token}" }
         extras = { auth_header: auth_header, expires: 1.hour.from_now, file_name: file.name, file_size: file.size.to_i }
         [download_url, extras]
       end
 
+      # @return [String]
+      # Authorization url that is used to request the initial access code from Box
       def auth_link
-        oauth_client.authorize_url(callback.to_s)
+        box_session.authorize_url(callback.to_s)
       end
 
+      # @return [Boolean]
       def authorized?
-        @token.present? && @token['token'].present?
+        box_token.present? && box_refresh_token.present? && !token_expired?
       end
 
+      # @return [Hash]
+      # Gets the appropriate tokens from Box using the access code returned from :auth_link:
       def connect(params, _data)
-        access_token = oauth_client.get_access_token(params[:code])
-        @token = { 'token' => access_token.token, 'refresh_token' => access_token.refresh_token }
+        register_access_token(box_session.get_access_token(params[:code]))
       end
 
       private
 
-        def oauth_client
+        def token_expired?
+          return true if expiration_time.nil?
+          Time.now.to_i > expiration_time
+        end
+
+        def box_client
+          if token_expired?
+            session = box_session(box_token)
+            register_access_token(session.refresh_token(box_refresh_token))
+          end
+          RubyBox::Client.new(box_session(box_token, box_refresh_token))
+        end
+
+        def box_session(token = nil, refresh_token = nil)
           RubyBox::Session.new(client_id: config[:client_id],
-                               client_secret: config[:client_secret])
-          # TODO: error checking here
+                               client_secret: config[:client_secret],
+                               access_token: token,
+                               refresh_token: refresh_token)
         end
 
-        def token_expired?(token)
-          return false unless @token.present? && @token['token'].present?
-          new_session = RubyBox::Session.new(
-            client_id: config[:client_id],
-            client_secret: config[:client_secret],
-            access_token: token
-          )
-          result = new_session.get("#{RubyBox::API_URL}/users/me")
-          result['status'] != 200
-        rescue RubyBox::AuthError => e
-          Rails.logger.error("AuthError occured when checking token. Exception #{e.class.name} : #{e.message}. token as expired and need to refresh it")
-          return true
+        # @param [OAuth2::AccessToken] access_token
+        def register_access_token(access_token)
+          @token = {
```

Comment:
@awead is there a reason you are not just toring the accesss_token and making a hash here?